### PR TITLE
Extend checkpoint loading to accept a validation function

### PR DESCRIPTION
**Pull request by irenedea**
# What does this PR do?
If a validation function is specified by the environment variable `CHECKPOINT_VALIDATION_FUNCTION`, the function is called prior to loading checkpoints via `torch.load`. 

The validation function is expected to be type `Callable[[Union[str, Path], bool]`, accepting the path to the checkpoint file and returning a boolean denoting whether or not the checkpoint is valid.

# What issue(s) does this change relate to?

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Comment by irenedea** [comment]
it's ready for review! @eracah @dakinggg

File: tests/trainer/test_checkpoint.py [hunk 1]
```diff
@@ -1289,3 +1290,42 @@ def test_rotate_checkpoints(
...
 
     dist.barrier()  # all ranks finish before cleaning up tmpdir
+
+
+def simple_validate(filepath: str):
+    with open(filepath, 'r') as f:
+        return f.read() == 'good'
+
+
+def test_checkpoint_validation(tmp_path):
```
Comment (dakinggg) [inline tests/trainer/test_checkpoint.py:1300]: would be good to test this integration with trainer too, in addition to these unit tests

**Review by dakinggg — APPROVED** [review]
This LGTM, can you please open a PR to foundry to make sure everywhere we call `torch.load` uses the `safe_torch_load` util from composer?

File: composer/utils/checkpoint.py [hunk 2]
```diff
@@ -39,6 +40,62 @@
...
+    fn = getattr(module, fn_name)
+    log.debug(f'Checkpoint validation function {name} has been found.')
+    return fn
+
+
+def _ensure_valid_checkpoint(checkpoint_filepath: Union[Path, str]) -> Union[Path, str]:
+    """Ensures that the checkpoint at checkpoint_filepath is valid.
+
+    using the function specified by the CHECKPOINT_VALIDATION_FUNCTION environment variable.
+    If CHECKPOINT_VALIDATION_FUNCTION, we skip validation.
```
[the remark this answers is not in the document]
Comment (irenedea) [reply composer/utils/checkpoint.py:66]:
```suggestion
    If CHECKPOINT_VALIDATION_FUNCTION is not set, we skip validation.
```

File: composer/utils/checkpoint.py [hunk 3]
```diff
@@ -716,7 +773,7 @@ def safe_torch_load(
 
             return state_dict
         else:
-            return torch.load(composer_states_filepath, map_location=map_location)
+            return _torch_load_with_validation(composer_states_filepath, map_location=map_location)
```
Comment (eracah) [inline composer/utils/checkpoint.py:776]: note that elastic sharded checkpoints don't use torch.load (they use `torch.distributed.checkpoint.load_state_dict`), so this validation function wouldn't be called for that

File: composer/utils/checkpoint.py [hunk 4]
```diff
@@ -39,6 +40,62 @@
...
+def _ensure_valid_checkpoint(checkpoint_filepath: Union[Path, str]) -> Union[Path, str]:
+    """Ensures that the checkpoint at checkpoint_filepath is valid.
+
+    using the function specified by the CHECKPOINT_VALIDATION_FUNCTION environment variable.
+    If CHECKPOINT_VALIDATION_FUNCTION is not set, we skip validation.
+
+    Args:
+        checkpoint_filepath (Union[Path,str]): The path to the checkpoint file.
+    """
+    fn_name = os.environ.get('CHECKPOINT_VALIDATION_FUNCTION', None)
```
Comment (eracah) [inline composer/utils/checkpoint.py:71]: why is the function is set by env variable? that seems very strange. why? and is there a bank of validation functions somewhere or something. are we setting the env variable or a user theoretically would? why would this not be a Trainer arg?

**Review by eracah — COMMENTED** [review]
Nice! Sleekly done! looks pretty good, but I just had a few questions and comments

File: composer/utils/checkpoint.py (hunk 3 above)
Comment (dakinggg) [reply composer/utils/checkpoint.py:776]: where is the file load for sharded checkpoints? Doesn't it still go through `_restore_checkpoint` (which starts by loading the state dict into memory)

Comment (dakinggg) [reply composer/utils/checkpoint.py:776]: ah nvm, i see it is `load_sharded_checkpoint`

Comment (dakinggg) [reply composer/utils/checkpoint.py:776]: Do you know where we can inject this for sharded checkpoints? It needs to run before we load the file into memory. Maybe loop over folder before calling `load_state_dict`?

File: composer/utils/checkpoint.py [hunk 5]
```diff
@@ -338,8 +399,35 @@ def _get_num_ranks_that_saved_rng(metadata: Metadata):
...
+        """FileSystemReader that validates checkpoint files prior to reading."""
+
+        def __init__(self, path: str):
+            super().__init__(path)
+
+        def read_data(self, plan: LoadPlan, planner: LoadPlanner):
+            """Reads metadata file.
+
+            Raises:
+                ValueError if the metadata file is invalid.
```
Comment (irenedea) [inline composer/utils/checkpoint.py:412]:
```suggestion
            """Reads data files.

            Raises:
                ValueError if the data files are not valid checkpoints.
```

File: composer/utils/checkpoint.py [hunk 6]
```diff
@@ -338,8 +399,37 @@ def _get_num_ranks_that_saved_rng(metadata: Metadata):
...
+            """
+            metadata_file_path = self.path / '.metadata'
+            _ensure_valid_checkpoint(metadata_file_path)
+            return super().read_metadata()
+
     # A subclass of FileSystemReader that downloads files from the object store before reading them from the local filesystem.
```
Comment (irenedea) [inline composer/utils/checkpoint.py:431]:
```suggestion
    # A subclass of FileSystemReaderWithValidation that downloads files from the object store before reading them from the local filesystem.
```

**Review by eracah — APPROVED** [review]
well done! LGTM

**Pull request merged**